### ISSUE TITLE
[Bug] If a modal dialog has an string array parameter, moving between text fields with the tab button resets their value

### ISSUE THREAD

**Issue by paleocomburo** [Blazored/Modal#508]
**Describe the bug**
I have a modal dialog that has a number of parameters that are being set for display. For example an edit dialog, where the parameters are the current values that the user can change.
If one of those parameters is a string array the behaviour of the modal changes. If there is no string array parameter present, and you update one of the values in a text field and then press the tab button to move to the next field, the values remain as expected. However, if you have a string array parameter on the dialog, then pressing the tab button will reset the text fields to their original value, losing the changes made by the user.

**To Reproduce**
1. Create a Blaxor WASM project. Add a new modal dialog.
2. Add 3 [Parameter] parameters: 2 string fields, and 1 string[] field. Show the two string fields als text boxes in a form.
3. Show the dialog, while providing values for the three parameters. Those values should show up on the dialog whe displayed.
4. Edit one of the text fields, then press the tab button to go to the next text field.
5. Notice how the value that was just changes has now reverted back to the original input value.

I have a repro project available here: https://github.com/paleocomburo/blazor-modal-refresh-repro

**Expected behavior**
Pressing the tab button should not revert the updated text field values.

**Hosting Model (is this issue happening with a certain hosting model?):**
 - Blazor WebAssembly
- ASP.NET Core hosted

**Additional context**
I have also tried using List<string>, ICollection<string> and IEnumeralbe<string>, all with the same result.
The behaviour only occurs then setting the array as a ModalParameter. If you have an array property on the dialog, but don't set it when showing the dialog, everything works. As soon as you provide an array as a ModalParameter input, the bug happens.

**Comment by paleocomburo** [comment]
Some additions, as I tried to find a work-around. It seems like every "complex" type causes this behaviour. Parameters of type `string`, `bool`, `int`, `float`, all work fine. But when I use `object`, or `Exception`, I also get the field reset behaviour.
I tried putting the array within a tuple, and a custom class, but all caused the problem.

**Comment by paleocomburo** [comment]
I tried to debug the issue, but I don't know enough about low-level Blazor to figure this out.
I did figure out it's related to the FocusTrap.
Disabling the focus trap via the global options:
``` html
<CascadingBlazoredModal ActivateFocusTrap="false">
```
does not work unfortunately.
The best way I have found so far to "fix" this, is by commenting out the focus trap altogether.
In BlazoredModalInstance.razor:
``` html
@if (UseCustomLayout)
{
    <CascadingValue Value="this" IsFixed="true">
        @Content
    </CascadingValue>
}
else
{
    <div class="bm-container @Position @OverlayCustomClass @OverlayAnimationClass"
         @ref="_modalReference" 
         @onclick="HandleBackgroundClick">

        @*<FocusTrap @ref="FocusTrap" IsActive="ActivateFocusTrap">*@
            <div class="@ModalClass" role="dialog" aria-modal="true" @onclick:stopPropagation="true">
                @if (!HideHeader)
                {
                    <div class="bm-header">
                        <h3 class="bm-title">@Title</h3>
                        @if (!HideCloseButton)
                        {
                            <button type="button" class="bm-close" aria-label="close" @onclick="() => CancelAsync()" @attributes="@_closeBtnAttributes"> 
                                <span>&times;</span>
                            </button>
                        }
                    </div>
                }
                <div class="bm-content">
                    <CascadingValue Value="this" IsFixed="true">
                        @Content
                    </CascadingValue>
                </div>
            </div>
        @*</FocusTrap>*@
        
    </div>
}
```

Or by disabling more logic via the IsActive boolean property on the FocusTrap itself:
``` html
@if (IsActive)
{
    <div class="blazored-modal-focus-trap" @ref="_container" @onkeydown="HandleKeyPresses" @onkeyup="HandleKeyPresses">
        <div tabindex="@(IsActive ? 0 : -1)" @ref="_startSecond" @onfocus="FocusEndAsync"></div>
        <div tabindex="@(IsActive ? 0 : -1)" @ref="_startFirst" @onfocus="FocusEndAsync"></div>
        @ChildContent
        <div tabindex="@(IsActive ? 0 : -1)" @ref="_endFirst" @onfocus="FocusStartAsync"></div>
        <div tabindex="@(IsActive ? 0 : -1)" @ref="_endSecond" @onfocus="FocusStartAsync"></div>
    </div>
}
else
{
    @ChildContent
}
```
And:
``` c#
    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (IsActive)
        {
            if (firstRender)
            {
                await _startFirst.FocusAsync();
            }
        }
    }
```

Which means that for now my work-around will be a custom version of this library. One where the focus trap is completely disabled.

**Comment by chrissainty** [comment]
Hi @paleocomburo. Sorry for not getting to this sooner. 

I've had a look at your repro and i see you're binding directly to the Parameters, this is a bad practice as Blazor will overwrite these values during rerenders. 

If you create local fields and assign the values from the Parameters to those fields, then bind to the local fields, I think you should see your desired outcome.

**Comment by paleocomburo** [comment]
I did just that. And it didn't work.

I created two new (private) local string fields. I bind them to the inputs. Then in the `OnParametersSetAsync` I set the values of these fields from the value of the parameter properties. Still same effect. If I press `tab` the values get reset.

The form:
``` html
<div class="m-2">
    <form>
        <p><b>Field A</b></p>
        <input type="text" size="75" class="form-control" @bind="_fieldA" />

        <p><b>Field B</b></p>
        <input type="text" size="75" class="form-control" @bind="_fieldB" />

        <button @onsubmit="() => Close()" @onsubmit:preventDefault @onclick="() => Close()" @onclick:preventDefault class="btn btn-primary">Ok</button>
        <button @onclick="() => BlazoredModal.CancelAsync()" @onclick:preventDefault class="btn btn-secondary">Cancel</button>
    </form>
</div>
```

The code:
``` c#
{
    [CascadingParameter]
    BlazoredModalInstance BlazoredModal { get; set; }

    [Parameter]
    public string FieldA { get; set; }


    [Parameter]
    public string FieldB { get; set; }


    [Parameter]
    public string[] ArrayA { get; set; }



    private string _fieldA;
    private string _fieldB;



    protected override Task OnParametersSetAsync()
    {
        _fieldA = FieldA;
        _fieldB = FieldB;

        return base.OnParametersSetAsync();
    }

    private async Task Close() => await BlazoredModal.CloseAsync();
}
```

Note that the property `ArrayA` is never used. It's only present as a parameter on the form, but it is not used in any binding. Just its present is enough to trigger the bug.

**Comment by chrissainty** [comment]
The issue is that you've assigned the field in the `OnParametersSet` method. Change to the `OnInitialized` lifecycle method and the issue is resolved.

**Comment by paleocomburo** [comment]
Well damn. I had actually tried that first, but the input fields were empty, so I figured it was because the parameters were not filled yet. But now I see I had missed the `return` in front of the call to `base.OnInitializedAsync()`.

I will try this tomorrow at work in the actual project and see if that solves it.
Thanks so far!

**Comment by chrissainty** [comment]
I'm glad you've made progress. You don't need the call to base.OnInitialized, I know some people do this out of habit but there is no code in the base method.

**Comment by paleocomburo** [comment]
Yeah, I do that because I was taught in my old C++ days that this is a good practice.
Anyway, I updated all my dialogs to use local fields for parameters that use binding, and now everything works as expected. Thank you very much for your time and advice.

You can close this defect as PEBCAK 😄

**Comment by chrissainty** [comment]
Glad you got it all sorted, @paleocomburo
